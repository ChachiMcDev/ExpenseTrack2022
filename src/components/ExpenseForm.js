import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'


const date = moment()
date.format('MMM Do, YYYY')
console.log(date.format('MMM Do, YYYY'))

//use local component to manage state
//upon submiting the form is when to add it to store/state



// constructor(props) {
//     super(props)
//     this.handleDeleteOptions = this.handleDeleteOptions.bind(this);
//     this.handleAddOption = this.handleAddOption.bind(this);
//     this.handlePick = this.handlePick.bind(this)
//     this.handleDeleteOne = this.handleDeleteOne.bind(this)

//     this.state = {
//         options: []
//     };

// };


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)


        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? parseFloat(props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ''
        }

    }

    //parseFloat(this.state.amount, 10) * 100
    // state = (props) => {
    //     if (props.editExpense) {
    //         return { ...editExpense }
    //     } else {
    //         return {
    //             description: '',
    //             note: '',
    //             amount: '',
    //             createdAt: moment(),
    //             calenderFocused: false,
    //             error: ''
    //         }
    //     }
    // }

    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }

    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }


    onAmountchange = (e) => {
        const amount = e.target.value;

        if (!amount || amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }

    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }))
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calenderFocused: focused }))
    }

    onFormSubmit = (e) => {
        e.preventDefault()
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and/or amount' }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }

    render() {
        return (
            <div>

                <div><h1>{this.state.error}</h1></div>
                <form
                    onSubmit={(e) => { this.onFormSubmit(e) }}
                >
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountchange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}>
                    </textarea>
                    <button>Add Expense</button>
                </form>

            </div>
        )
    }
}


