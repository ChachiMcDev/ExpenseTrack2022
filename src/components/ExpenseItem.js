import React from 'react'
import { connect } from 'react-redux'
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom'




export const ExpenseItem = ({ id, description, amount, note, createdAt, dispatch }) => (
    <div>
        <Link to={`/edit/${id}`}><h3>{description}</h3></Link>

        <p>{amount} - {createdAt} </p>
        <p>{note}</p>
        <button onClick={() => {
            dispatch(removeExpense({ id }))
        }}>Remove</button>

    </div>
)



export default connect()(ExpenseItem)




