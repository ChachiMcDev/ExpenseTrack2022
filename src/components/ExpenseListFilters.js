import React from 'react'
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates'
import { setTextFilter, setSortByDate, setSortByAmount, setStartDate, setEndDate } from '../actions/filters'





export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            calendarFocused: null
        }
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState({ calendarFocused })
    }


    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }

    onSortByChange = (e) => {
        const eventValue = e.target.value;
        eventValue === 'amount' ? this.props.setSortByAmount() : this.props.setSortByDate()

    }

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />

                <select value={this.props.filters.sortBy} onChange={this.onSortByChange}>
                    <option value="date">Date</option>
                    <option value="amount" >Amount</option>
                </select>

                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>

        )
    }

}


// const handleDropDown = (e, props) => {
//     const eventValue = e.target.value;
//     // let sortBy = props.filters.sortBy;
//     // eventValue === 'date' ? sortBy = 'date' : sortBy = 'amount'
//     eventValue === 'amount' ? props.dispatch(setSortByAmount()) : props.dispatch(setSortByDate())
// }

// const ExpenseListFilters = (props) => (

//     <div>
//         <input type="text" value={props.filters.text} onChange={(e) => {
//             console.log('heheheh', { ...props })
//             props.dispatch(setTextFilter(e.target.value))
//         }} />

//         <select value={props.filters.sortBy} onChange={(e) => {
//             handleDropDown(e, props)
//         }}>
//             <option value="date">Date</option>
//             <option value="amount" >Amount</option>
//         </select>

//         <DateRangePicker
//             startDate={props.filters.startDate}
//             endDate={props.filters.endDate}
//             onDatesChange={({ startDate, endDate }) => ({ filters: { startDate, endDate } })}

//         />
//     </div>
// )




const mapStateToProps = (state) => {

    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate)),
        setSortByAmount: () => dispatch(setSortByAmount()),
        setSortByDate: () => dispatch(setSortByDate())
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)


// const eventValue = e.target.value;
// let sortBy = props.filters.sortBy;
// eventValue === 'date' ? sortBy = 'date' : sortBy = 'amount'
// eventValue === 'amount' ? props.dispatch(setSortByAmount()) : props.dispatch(setSortByDate())

