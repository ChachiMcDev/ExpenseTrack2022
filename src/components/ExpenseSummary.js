import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectTotalExpenses from '../selectors/expenses-total';


export const ExpenseSummary = ({ expenseCount, totalExpenses }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpenseTotal = numeral(totalExpenses / 100).format('$0,0.00');

    return (
        <div>
            <h1>Viewing {expenseCount} {expenseWord} totalling {formattedExpenseTotal} </h1>
        </div>
    )
}


const mapStateToProps = (state) => {
    return {
        expenseCount: selectExpenses(state.expenses, state.filters).length,
        totalExpenses: selectTotalExpenses(selectExpenses(state.expenses, state.filters))
    };
};


export default connect(mapStateToProps)(ExpenseSummary)