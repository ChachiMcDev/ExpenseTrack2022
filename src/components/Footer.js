import React from 'react';


const Footer = () => (
    <header>
        <h1>Im the footer</h1>
        <div className="menu">social media and address and what not</div>
    </header>
)


export default Footer