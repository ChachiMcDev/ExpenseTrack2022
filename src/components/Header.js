import React from 'react';
import { connect } from 'react-redux';
import MenuLinks from './MenuLinks'



export const Header = ({ isAuthenticated }) => (
    <header>
        <h1>Expensify Header</h1>


        <div className="menu">Menu on the right</div>





        {isAuthenticated ? <MenuLinks /> : ""}


    </header>
)

const mapStateToProps = (state) => ({
    isAuthenticated: state.auth.uid
})


export default connect(mapStateToProps)(Header)




