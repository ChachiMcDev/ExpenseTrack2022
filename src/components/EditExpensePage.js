import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';




export class EditExpensePage extends React.Component {


    onSubmit = (expense) => {

        this.props.startEditExpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }

    onRemove = () => {

        this.props.startRemoveExpense({ id: this.props.expense.id })
        this.props.history.push('/')
    }

    render() {

        return (
            <div>

                <ExpenseForm

                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />

                <button
                    onClick={this.onRemove}
                >
                    Remove
                </button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    };
};

const mapDispatchToProps = (dispatch) => {


    return {
        startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
        startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
    }
}


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);


// const EditExpensePage = (props) => {
//     // console.log('thesePROPS', props)

//     // const id = props.match.params.id
//     // console.log(id)
//     // // const getExp = props.expenses.find((expense) => {
//     // //     return id === expense.id
//     // // })

//     // const editExp = props.expenses.filter((ex) => {
//     //     return ex.id === id
//     // })
//     // console.log('expense with id', editExp)
//     const id = props.match.params.id
//     return (

//         <div>
//             editing id of {props.match.params.id}

//             <ExpenseForm
//                 expense={props.expense}
//                 onSubmit={(expense) => {
//                     props.dispatch(editExpense(id, expense))
//                     props.history.push('/')
//                 }} />

//             <button onClick={() => {
//                 props.dispatch(removeExpense({ id }))
//                 props.history.push('/')
//             }}>Remove</button>

//         </div>
//     )
// }


