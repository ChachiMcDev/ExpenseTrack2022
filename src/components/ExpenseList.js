import React from 'react';
import { connect } from 'react-redux';
import ExpenseItem from './ExpenseItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = ({ expenses }) => (
    <div>
        {
            expenses.length === 0 ? (<p>No Expenses</p>) : (expenses.map((ex) => (
                <ExpenseItem key={ex.id} {...ex} />
            )))
        }

    </div>
);


// export const ExpenseList = (props) => (
//     <div>

//         <h1>Expnese list</h1>
//         {props.expenses.map((y) => {
//             console.log(y)
//         })}
//         {props.expenses.map((x) => (

//             <ExpenseItem key={x.id} {...x} />
//         ))}
//     </div>
// )
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);

// description = { ex.description }
// amount = { ex.amount }
// note = { ex.note }
// createdAt = { ex.createdAt }