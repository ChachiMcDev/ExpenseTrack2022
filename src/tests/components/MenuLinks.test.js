import React from 'react';
import { shallow } from 'enzyme';
import { MenuLinks } from '../../components/MenuLinks';


test('should render menu links page correctly', () => {
    const wrapper = shallow(<MenuLinks />);
    expect(wrapper).toMatchSnapshot();

})



test('should call startLogout on link click', () => {
    const startLogout = jest.fn()
    const wrapper = shallow(<MenuLinks startLogout={startLogout} />);

    wrapper.find('#logout').simulate('click')
    expect(startLogout).toHaveBeenCalled();

})