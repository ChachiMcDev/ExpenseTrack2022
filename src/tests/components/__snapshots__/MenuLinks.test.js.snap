// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render menu links page correctly 1`] = `
<div>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    exact={true}
    to="/dashboard"
  >
    Dash
  </NavLink>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    to="/create"
  >
    Create
  </NavLink>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    id="logout"
    to="/"
  >
    LogOut
  </NavLink>
</div>
`;
